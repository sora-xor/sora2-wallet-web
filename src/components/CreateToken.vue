<template>
  <wallet-base :title="t('createToken.title')" show-back @back="handleBack">
    <div class="wallet-settings-create-token">
      <template v-if="step === Step.Create">
        <s-input
          :placeholder="t(`createToken.tokenSymbol.placeholder`)"
          :minlength="1"
          :maxlength="7"
          :disabled="loading"
          v-maska="tokenSymbolMask"
          v-model="tokenSymbol"
        />
        <p class="wallet-settings-create-token_desc">{{ t(`createToken.tokenSymbol.desc`) }}</p>
        <s-input
          :placeholder="t(`createToken.tokenName.placeholder`)"
          :minlength="1"
          :maxlength="33"
          :disabled="loading"
          v-maska="tokenNameMask"
          v-model="tokenName"
        />
        <p class="wallet-settings-create-token_desc">{{ t(`createToken.tokenName.desc`) }}</p>
        <s-float-input
          :placeholder="t(`createToken.tokenSupply.placeholder`)"
          v-model="tokenSupply"
          :decimals="decimals"
          :max="maxTotalSupply"
          :disabled="loading"
        />
        <p class="wallet-settings-create-token_desc">{{ t(`createToken.tokenSupply.desc`) }}</p>
        <div class="wallet-settings-create-token_supply-block">
          <span>{{ t(`createToken.extensibleSupply.placeholder`) }}</span>
          <s-switch v-model="extensibleSupply" :disabled="loading" />
        </div>
        <s-button
          class="wallet-settings-create-token_action"
          type="primary"
          :loading="loading"
          :disabled="!(tokenSymbol && tokenName.trim() && tokenSupply)"
          @click="onConfirm"
        >
          <template v-if="!tokenSymbol">{{ t('createToken.enterSymbol') }}</template>
          <template v-else-if="!tokenSupply">{{ t('createToken.enterSupply') }}</template>
          <template v-else>{{ t('createToken.action') }}</template>
        </s-button>
      </template>
      <template v-else-if="step === Step.Confirm">
        <div class="wallet-settings-create-token_confirm-block">
          <span>{{ t('createToken.tokenSymbol.placeholder') }}</span>
          <span>{{ tokenSymbol }}</span>
        </div>
        <div class="wallet-settings-create-token_confirm-block">
          <span>{{ t('createToken.tokenName.placeholder') }}</span>
          <span>{{ tokenName.trim() }}</span>
        </div>
        <div class="wallet-settings-create-token_confirm-block">
          <span>{{ t('createToken.tokenSupply.placeholder') }}</span>
          <span>{{ formattedTokenSupply }}</span>
        </div>
        <div class="wallet-settings-create-token_confirm-block">
          <span>{{ t('createToken.extensibleSupply.placeholder') }}</span>
          <span>{{ extensibleSupply ? 'Yes' : 'No' }}</span>
        </div>
        <s-divider class="wallet-settings-create-token_divider" />
        <div class="wallet-settings-create-token_confirm-block">
          <div class="wallet-settings-create-token_fee-block">
            <s-tooltip class="bridge-info-icon" popper-class="info-tooltip info-tooltip--bridge" border-radius="mini" :content="t('createToken.tooltipValue')" theme="light" placement="right-start" animation="none" :show-arrow="false">
              <s-icon name="info-16" />
            </s-tooltip>
            <span class="wallet-settings-create-token_fee-block_title">{{ t('createToken.fee') }}</span>
          </div>
          <span>{{ formattedFee }} {{ KnownSymbols.XOR }}</span>
        </div>
        <s-button
          class="wallet-settings-create-token_action"
          type="primary"
          :disabled="!hasEnoughXor"
          :loading="loading"
          @click="onCreate"
        >
          <template v-if="!hasEnoughXor">{{ t('createToken.insufficientBalance', { symbol: KnownSymbols.XOR }) }}</template>
          <template v-else>{{ t('createToken.confirm') }}</template>
        </s-button>
      </template>
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator'
import { Action } from 'vuex-class'
import { KnownSymbols, CodecString, KnownAssets, FPNumber, MaxTotalSupply } from '@sora-substrate/util'

import TransactionMixin from './mixins/TransactionMixin'
import WalletBase from './WalletBase.vue'
import { RouteNames } from '../consts'
import { api } from '../api'

enum Step {
  Create,
  Confirm
}

@Component({
  components: {
    WalletBase
  }
})
export default class CreateToken extends Mixins(TransactionMixin) {
  readonly KnownSymbols = KnownSymbols
  readonly Step = Step
  readonly decimals = FPNumber.DEFAULT_PRECISION
  readonly maxTotalSupply = MaxTotalSupply
  readonly tokenSymbolMask = 'AAAAAAA'
  readonly tokenNameMask = { mask: 'Z*', tokens: { Z: { pattern: /[0-9a-zA-Z ]/ } } }

  step = Step.Create
  tokenSymbol = ''
  tokenName = ''
  tokenSupply = ''
  extensibleSupply = false
  fee: CodecString = '0'

  @Action navigate

  handleBack (): void {
    if (this.step === Step.Create) {
      this.navigate({ name: RouteNames.Wallet })
    } else {
      this.step = Step.Create
    }
  }

  get formattedFee (): string {
    return this.formatCodecNumber(this.fee)
  }

  get formattedTokenSupply (): string {
    return this.formatStringValue(this.tokenSupply, this.decimals)
  }

  get hasEnoughXor (): boolean {
    const xor = KnownAssets.get(KnownSymbols.XOR)
    const accountXor = api.accountAssets.find(asset => asset.address === xor.address)
    if (!accountXor || !accountXor.balance || !+accountXor.balance.transferable) {
      return false
    }
    const fpAccountXor = this.getFPNumberFromCodec(accountXor.balance.transferable, accountXor.decimals)
    return FPNumber.gte(fpAccountXor, this.getFPNumberFromCodec(this.fee))
  }

  async calculateFee (): Promise<CodecString> {
    return api.getRegisterAssetNetworkFee(
      this.tokenSymbol,
      this.tokenName,
      this.tokenSupply,
      this.extensibleSupply
    )
  }

  async registerAsset (): Promise<void> {
    return api.registerAsset(
      this.tokenSymbol,
      this.tokenName.trim(),
      this.tokenSupply,
      this.extensibleSupply
    )
  }

  async onConfirm (): Promise<void> {
    if (!this.tokenSymbol.length || !this.tokenSupply.length) {
      return
    }
    const tokenSupply = this.getFPNumber(this.tokenSupply, this.decimals)
    const maxTokenSupply = this.getFPNumber(MaxTotalSupply, this.decimals)
    if (FPNumber.gt(tokenSupply, maxTokenSupply)) {
      this.tokenSupply = maxTokenSupply.toString()
    }
    try {
      await this.withLoading(async () => {
        this.fee = await this.calculateFee()
        this.step = Step.Confirm
      })
    } catch (error) {
      console.error(error)
      this.$notify({
        message: this.t('createToken.feeError'),
        type: 'error',
        title: ''
      })
    }
  }

  async onCreate (): Promise<void> {
    await this.withNotifications(
      async () => {
        if (!this.hasEnoughXor) {
          throw new Error('walletSend.badAmount')
        }
        await this.registerAsset()
        this.navigate({ name: RouteNames.Wallet })
      }
    )
  }
}
</script>

<style scoped lang="scss">
.wallet-settings-create-token {
  &_desc {
    color: var(--s-color-base-content-secondary);
    font-size: $font-size_small;
    padding: $basic-spacing_mini 0 $basic-spacing 0;
    line-height: var(--s-line-height-big);
    font-feature-settings: var(--s-font-feature-settings-common);
  }
  &_supply-block {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    background: var(--s-color-base-background);
    padding: $basic-spacing_small $basic-spacing;
    border-radius: $basic-spacing_mini;
    margin-bottom: $basic-spacing;
  }
  &_action {
    width: 100%;
  }
  &_confirm-block {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    color: var(--s-color-base-content-secondary);
    font-size: $font-size_basic;
    font-weight: normal;
    padding: $basic-spacing_mini 0;
    line-height: var(--s-line-height-big);
    font-feature-settings: var(--s-font-feature-settings-common);
  }
  &_fee-block {
    display: flex;
    flex-direction: row;
    align-items: center;
  }
  &_fee-block_title {
    margin-left: $basic-spacing_mini;
    line-height: 1;
  }
  &_divider {
    margin: unset;
  }
}
</style>
