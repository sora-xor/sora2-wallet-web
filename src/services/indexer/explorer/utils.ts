import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
