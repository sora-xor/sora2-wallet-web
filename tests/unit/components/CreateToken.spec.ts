import Vuex from 'vuex';
import { shallowMount } from '@vue/test-utils';

import { useDescribe, localVue } from '../../utils';
import { MOCK_CREATE_TOKEN } from '../../utils/CreateTokenMock';

import CreateToken from '@/components/CreateToken.vue';
import { MOCK_NETWORK_FEES } from '../../utils/NetworkFeesMock';

const createStore = () =>
  new Vuex.Store({
    modules: {
      Settings: {
        getters: {
          networkFees: () => MOCK_NETWORK_FEES,
        },
      },
      Router: {
        actions: {
          navigate: jest.fn(),
        },
      },
    } as any,
  });

useDescribe('CreateToken.vue', CreateToken, () => {
  MOCK_CREATE_TOKEN.map((item) =>
    it(`[${item.title}]: should be rendered correctly`, () => {
      const wrapper = shallowMount(CreateToken, {
        localVue,
        store: createStore(),
        data: () => {
          return {
            step: item.step,
            tokenSymbol: item.tokenSymbol,
            tokenName: item.tokenName,
            tokenSupply: item.tokenSupply,
            extensibleSupply: item.extensibleSupply,
          };
        },
        computed: {
          hasEnoughXor: () => item.hasEnoughXor,
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    })
  );
});
