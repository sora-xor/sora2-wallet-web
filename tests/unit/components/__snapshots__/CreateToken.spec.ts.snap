// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateToken.vue [Confirm Step]: should be rendered correctly 1`] = `
<wallet-base-stub
  showback="true"
  showheader="true"
  title="Create token"
  tooltip=""
>
  <div
    class="wallet-settings-create-token"
  >
    <info-line-stub
      assetsymbol=""
      fiatvalue=""
      label="Token symbol"
      labeltooltip=""
      value="TSYMBOL"
    />
     
    <info-line-stub
      assetsymbol=""
      fiatvalue=""
      label="Token name"
      labeltooltip=""
      value="Token name"
    />
     
    <info-line-stub
      assetsymbol=""
      fiatvalue=""
      label="Token supply"
      labeltooltip=""
      value="10,000,000,000"
    />
     
    <info-line-stub
      assetsymbol=""
      fiatvalue=""
      label="Extensible supply"
      labeltooltip=""
      value="Yes"
    />
     
    <s-button-stub
      borderradius="small"
      class="wallet-settings-create-token_action s-typography-button--large"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
      usedesignsystem="true"
    >
      Confirm
    </s-button-stub>
     
    <wallet-fee-stub
      value="0.007"
    />
  </div>
</wallet-base-stub>
`;

exports[`CreateToken.vue [Create Step]: should be rendered correctly 1`] = `
<wallet-base-stub
  showback="true"
  showheader="true"
  title="Create token"
  tooltip=""
>
  <div
    class="wallet-settings-create-token"
  >
    <s-input-stub
      accept=""
      autocomplete="off"
      borderradius="small"
      form=""
      label=""
      maxlength="7"
      minlength="1"
      name=""
      placeholder="Token symbol"
      prefix=""
      size="big"
      suffix=""
      tabindex=""
      type="text"
      usedesignsystem="true"
      value=""
    />
     
    <p
      class="wallet-settings-create-token_desc"
    >
      This is an identifier for the token. You can use from 1 to 7 characters for the selection.
    </p>
     
    <s-input-stub
      accept=""
      autocomplete="off"
      borderradius="small"
      form=""
      label=""
      maxlength="33"
      minlength="1"
      name=""
      placeholder="Token name"
      prefix=""
      size="big"
      suffix=""
      tabindex=""
      type="text"
      usedesignsystem="true"
      value=""
    />
     
    <p
      class="wallet-settings-create-token_desc"
    >
      This is a name for the token. You can use from 1 to 33 characters for the selection.
    </p>
     
    <s-float-input-stub
      decimals="18"
      delimiters="[object Object]"
      haslocalestring="true"
      max="170141183460469231731.687303715884105727"
      placeholder="Token supply"
      value=""
    />
     
    <p
      class="wallet-settings-create-token_desc"
    >
      You will receive these tokens upon creation.
    </p>
     
    <div
      class="wallet-settings-create-token_supply-block"
    >
      <s-switch-stub
        activeiconСlass=""
        activetext=""
        activevalue="true"
        inactiveiconСlass=""
        inactivetext=""
        name=""
        usedesignsystem="true"
        width="40"
      />
       
      <span>
        Extensible supply
      </span>
    </div>
     
    <p
      class="wallet-settings-create-token_desc"
    >
      Ability to mint more tokens.
    </p>
     
    <s-button-stub
      borderradius="small"
      class="wallet-settings-create-token_action s-typography-button--large"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
      usedesignsystem="true"
    >
      Enter token symbol
    </s-button-stub>
     
    <!---->
  </div>
</wallet-base-stub>
`;
